src/ocaml_monadic_ppx.ml

```diff
@@ -1,113 +1,130 @@
+        ]
+      in
+      let f = Ast_helper.Exp.function_ cases in
+      ([%expr bind [%e expr_if] [%e f]] [@metaloc expr.pexp_loc])
+  | Pexp_sequence (expr_seq_l, expr_seq_r) ->
+      [%expr bind [%e expr_seq_l] (fun () -> [%e expr_seq_r])]
+      [@metaloc expr.pexp_loc]
+  | _ -> expr
+
+let orzero_expand ~ctxt expr =
+  let loc = Expansion_context.Extension.extension_point_loc ctxt in
+  match expr.pexp_desc with
+  | Pexp_let (Nonrecursive, value_bindings, body) ->
+      (* This is a let%orzero expression.  It's of the form
+           let%orzero $p1 = $e1 and ... and $pn = $en in $e0
+         and we want it to take the form
+           match $e1 with
+           | $p1 -> (match $e2 with
+                     | $p2 -> ...
+                              (match $en with
+                               | $pn -> $e0
+                               | _ -> zero ())
+                     | _ -> zero ())
+           | _ -> zero ()
+      *)
+      let rec orzero_wrap value_bindings' =
+        match value_bindings' with
+        | {
+            pvb_pat = orzero_pattern;
+            pvb_expr = orzero_expr;
+            pvb_attributes = [];
+            pvb_loc = _orzero_loc;
+          }
+          :: value_bindings'' ->
+            (* Recurse and then wrap the resulting body. *)
+            let body' = orzero_wrap value_bindings'' in
+            ([%expr
+               match [%e orzero_expr] with
+               | [%p orzero_pattern] -> [%e body']
+               | _ -> zero ()]
+            [@metaloc expr.pexp_loc])
+        | _ ->
+            (* Nothing left to do.  Just return the body. *)
+            body
+      in
+      orzero_wrap value_bindings
+  | _ -> expr
+
+let bind_extension =
+  Extension.V3.declare "bind" Extension.Context.expression
+    Ast_pattern.(single_expr_payload __)
+    bind_expand
+
+let orzero_extension =
+  Extension.V3.declare "orzero" Extension.Context.expression
+    Ast_pattern.(single_expr_payload __)
+    orzero_expand
+
+let orzero_rule = Ppxlib.Context_free.Rule.extension orzero_extension
+
+let bind_rule = Ppxlib.Context_free.Rule.extension bind_extension
+
+let expr_mapper =
+  object
+    inherit Ast_traverse.map as super
+
+    method! expression e =
+      let e = super#expression e in
+      match e with
+      | [%expr
+          [%guard [%e? guard_expr]];
+          [%e? body_expr]] ->
+          (* This is a sequenced expression with a [%guard ...] extension.  It
+             takes the form
+               [%guard expr']; expr
+             and we want it to take the form
+               if expr' then expr else zero ()
+          *)
+          let loc = e.pexp_loc in
           [%expr if [%e guard_expr] then [%e body_expr] else zero ()]
-          [@metaloc outer_expr.pexp_loc]
-      | _ -> default_mapper.expr mapper outer_expr
-  }
-;;
+      | _ -> e
+  end
+
+let () =
+  Driver.register_transformation "ocaml-monadic"
+    ~rules:[ bind_rule; orzero_rule ] ~impl:expr_mapper#structure
```

Comment:
This is really just a side-comment, but I thought it would be good to at least mention it: registering a ppx as a whole file transformation (as done here via the `~impl` argument) should only be done as a last resource. That's because for every registered whole file transformation, the AST has to be traversed one whole more time. So due to this `%guard` extension point, using `ocaml-monadic` means traversing the AST one more time when compiling (even if `%guard` is never used).

But there's no way to do this better: the `%guard` syntax doesn't allow to use a context-free rule (as done for the other extension points here). For a context-free rule, the syntax would need to be chosen in a way that everything that gets expanded forms part of the payload; something like
```
[%guard b;
return x]
```
instead of
```
[%guard b];
return x
```

Reply:
Thanks for the explanation @pitag-ha 😊 I first tried to implement it using a context-free rule, but as you pointed out, that was not possible with the current syntax.